solidity/contracts/converter/types/liquidity-pool-v3/LiquidityPoolV3Converter.sol

```diff
@@ -0,0 +1,1356 @@
+    */
+    function convert(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, address _trader, address payable _beneficiary)
+        public
+        override
+        payable
+        protected
+        only(BANCOR_NETWORK)
+        returns (uint256)
+    {
+        // validate input
+        require(_sourceToken != _targetToken, "ERR_SAME_SOURCE_TARGET");
+
+        // if a whitelist is set, verify that both and trader and the beneficiary are whitelisted
+        require(address(conversionWhitelist) == address(0) ||
+                (conversionWhitelist.isWhitelisted(_trader) && conversionWhitelist.isWhitelisted(_beneficiary)),
+                "ERR_NOT_WHITELISTED");
+
+        return doConvert(_sourceToken, _targetToken, _amount, _trader, _beneficiary);
+    }
+
+    /**
+      * @dev returns the conversion fee for a given target amount
+      *
+      * @param _targetAmount  target amount
+      *
+      * @return conversion fee
+    */
+    function calculateFee(uint256 _targetAmount) internal view returns (uint256) {
+        return _targetAmount.mul(conversionFee) / PPM_RESOLUTION;
+    }
+
+    /**
+      * @dev gets the stored reserve balance for a given reserve id
+      *
+      * @param _reserveId   reserve id
+    */
+    function getReserveBalance(uint256 _reserveId) internal view returns (uint256) {
+        return (reserveBalances >> ((_reserveId - 1) * 128)) & MAX_UINT128;        
+    }
+
+    /**
+      * @dev gets the stored reserve balances
+      *
+      * @param _sourceId    source reserve id
+      * @param _targetId    target reserve id
+    */
+    function getReserveBalances(uint256 _sourceId, uint256 _targetId) internal view returns (uint256, uint256) {
+        require(_sourceId + _targetId == 3, "ERR_INVALID_RESERVES");
+        uint256 reserveBalancesLocal = reserveBalances;
+        return (
+            (reserveBalancesLocal >> ((_sourceId - 1) * 128)) & MAX_UINT128,
+            (reserveBalancesLocal >> ((_targetId - 1) * 128)) & MAX_UINT128
+        );        
+    }
+
+    /**
+      * @dev sets the stored reserve balance for a given reserve id
+      *
+      * @param _reserveId       reserve id
+      * @param _reserveBalance  reserve balance
+    */
+    function setReserveBalance(uint256 _reserveId, uint256 _reserveBalance) internal {
+        require(_reserveBalance <= MAX_UINT128, "ERR_RESERVE_BALANCE_OVERFLOW"); 
+        uint256 otherBalance = getReserveBalance(3 - _reserveId);
+        reserveBalances = (_reserveBalance << ((_reserveId - 1) * 128)) | (otherBalance << ((2 - _reserveId) * 128));
+    }
+
+    /**
+      * @dev syncs the stored reserve balance for a given reserve with the real reserve balance
+      *
+      * @param _reserveToken    address of the reserve token
+    */
+    function syncReserveBalance(IERC20Token _reserveToken) internal {
+        uint256 reserveId = reserveIds[_reserveToken];
```

Comment:
we could probably save a storage read here if the function receives the id instead of the address in some cases

Reply:
This function is never used during conversion.

Reply:
yeah, that's for other flows. maybe just an overload.